docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

```diff
@@ -0,0 +1,1341 @@
+.. Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+.. _upgrade_model_reference:
+
+The Upgrade Model in Depth - Reference
+======================================
+
+This section describes in detail what rules govern package validation
+upon upload and how contracts are upgraded or downgraded at runtime.
+These topics are for a large part covered in the sections above. This
+section acts as a thorough reference.
+
+Static Checks
+-------------
+
+Upgrade static checks are performed once alongside other validity checks
+when a DAR is uploaded to a participant. DARs deemed invalid for
+upgrades are rejected.
+
+DARs upgrade checks are broken down into package-level checks, in turn
+broken down into template and data type-level checks.
+
+Packages
+~~~~~~~~
+
+**Definition:** A *utility package* is a package with no template
+definition, no interface definition, and only
+non-`serializable <https://github.com/digital-asset/daml/blob/main-2.x/sdk/daml-lf/spec/daml-lf-1.rst#serializable-types>`__
+data type definitions. A utility package is typically only made of
+helper functions and constants.
+
+In the following, packages whose LF version does not support upgrades
+(1.15 and earlier) and utility packages are ignored. Everything is as if
+they didn’t exist for the purposes of the package validity check.
+
+Upon upload to a participant, a DAR is checked against previously
+uploaded DARs for upgrade validity. More precisely, for every package
+with name *p* and version *v* present in the uploaded dar:
+
+1. The participant looks up versions *v_prev* and *v_next* of *p* in its package database, such that *v_prev* is the greatest version of
+   *p* smaller than *v*, and *v_next* is the smallest version of *p*
+   greater than *v.*
+
+   .. image:: ./images/upgrade-static-checks-order.svg
+
+2. The participant checks that version *v* of *p* is a valid upgrade of
+   version *v_prev* of p.
+
+3. The participant checks that version *v_next* of *p* is a valid
+   upgrade of version *v* of *p*.
+
+| This means that “being a valid upgrade” for a package is context
+  dependent: it depends on what packages are already installed on the
+  participant.
+| It is also *modular*: checks are performed at the package level. That
+  is, a new version of a package is rejected as soon as it contains some
+  element which doesn’t properly upgrade its counterpart in the old
+  package, even if some other elements do.
+
+Modules
+~~~~~~~
+
+The modules of the upgrading package must form a superset of the modules
+of the upgraded package. In other words, it is valid to add new modules
+but deleting a module leads to a validation error.
+
+Examples
```

Comment:
For the docs team - is it possible to keep these "Examples" as "titles", i.e. slightly bigger and bold, but not show them in the TOC, as the TOC for this page is very polluted with the word "Examples"

Reply:
Fair enough, @paulbrauner-da could you make these changes?

Reply:
Sure, will do. Actually Carrie did most of them already.